tools/kconfiglib: Add riot_kconfig to override default behaviours

leandrolanzieri opened the pull request:
### Contribution description
This PR adds a `RiotKconfig` class which can be used to override behaviours from the base `Kconfig`. It also patches menuconfig.py to use the new class. For now it implements a version of `_parse_help` which removes Doxygen markers.

`standard_riot_kconfig` is taken from the original `standard_kconfig`.

### Testing procedure
- Check that `make menuconfig` still works as usual
- Check that the help string of a symbol with Doxygen markers (e.g. `GNRC_NETIF_IPV6_ADDRS_NUMOF`) does not show them.


### Issues/PRs references
Solves issue presented here https://github.com/RIOT-OS/RIOT/pull/13227#discussion_r412118812

bergzand commented:
@aabadie Do you mind give this your usual python quality check? :)

aabadie commented on dist/tools/kconfiglib/riot_kconfig.py:
```diff
@@ -0,0 +1,54 @@
+    """ RIOT adaption of Kconfig class """
+    def _parse_help(self, node):
+        """ Parses the help section of a node, removing Doxygen markers """
+        DOXYGEN_MARKERS = ["@ref ", "@see "]
```
```suggestion
        doxygen_markers = ["@ref ", "@see "]
```

leandrolanzieri replied:
Changed

aabadie commented on dist/tools/kconfiglib/riot_kconfig.py:
```diff
@@ -0,0 +1,54 @@
+        super(RiotKconfig, self)._parse_help(node)
+
+        # remove Doxygen markers
+        for marker in DOXYGEN_MARKERS:
```
```suggestion
        for marker in doxygen_markers:
```

aabadie commented on dist/tools/kconfiglib/riot_kconfig.py:
```diff
@@ -0,0 +1,54 @@
+      The 'description' passed to argparse.ArgumentParser().
+      argparse.RawDescriptionHelpFormatter is used, so formatting is preserved.
+    """
+    import argparse
```
Should be added at the beginning of the module

leandrolanzieri replied:
Moved

leandrolanzieri commented:
> Just minor comments for the moment.
> 
> Do you think it would be possible to monkey patch the `_main` function of the menuconfig.py module in order to use `standard_riot_kconfig` instead of patching the file itself ?

Thanks for the suggestion. I added a `riot_menuconfig.py` which calls the `menuconfig` function now with the RIOT-specific class.

> In fact, I'm wondering why add kconfiglib as a package instead of asking it to be installed using pip (a bit off-topic in this PR) ?

I added it as a package because it appeared to me as a ready-to-run solution, instead of burdening the user with installing an extra tool and adding another dependency.

aabadie commented:
>  I added a riot_menuconfig.py which calls the menuconfig function now with the RIOT-specific class.

Thanks, I prefer this approach!

> I added it as a package because it appeared to me as a ready-to-run solution, instead of burdening the user with installing an extra tool and adding another dependency.

Ok

aabadie commented:
I'm testing this PR but I don't understand what I'm supposed to get with "Check that the help string of a symbol with Doxygen markers (e.g. GNRC_NETIF_IPV6_ADDRS_NUMOF) does not show them."

Any hint @leandrolanzieri ?

leandrolanzieri commented:
> I'm testing this PR but I don't understand what I'm supposed to get with "Check that the help string of a symbol with Doxygen markers (e.g. GNRC_NETIF_IPV6_ADDRS_NUMOF) does not show them."
> 
> Any hint @leandrolanzieri ?

Yeah sorry. What I mean is that in master you should see in the menuconfig interface the help strings with Doxygen markers, and with this PR those should not appear. The help string can be displayed either toggling the 'show-help' mode (`F`) or showing the symbol info (`?`) when the symbol is highlighted.

- For `GNRC_NETIF_IPV6_ADDRS_NUMOF`:
**master**:
```
Name: GNRC_NETIF_IPV6_ADDRS_NUMOF                                                                                                                               
Prompt: Maximum number of unicast and anycast addresses per interface                                                                                           
Type: int                                                                                                                                                       
Value:                                                                                                                                                          
                                                                                                                                                                
Help:                                                                                                                                                           
                                                                                                                                                                
  If you change this, please make sure that                                                                                                                     
  @ref CONFIG_GNRC_NETIF_IPV6_GROUPS_NUMOF is also large enough to fit the                                                                                      
  addresses' solicited nodes multicast addresses.                                                                                                               
  Default: 2 (1 link-local + 1 global address).     
```
**This PR**
```
Name: GNRC_NETIF_IPV6_ADDRS_NUMOF                                                                                                                               
Prompt: Maximum number of unicast and anycast addresses per interface                                                                                           
Type: int                                                                                                                                                       
Value:                                                                                                                                                          
                                                                                                                                                                
Help:                                                                                                                                                           
                                                                                                                                                                
  If you change this, please make sure that                                                                                                                     
  CONFIG_GNRC_NETIF_IPV6_GROUPS_NUMOF is also large enough to fit the                                                                                           
  addresses' solicited nodes multicast addresses.                                                                                                               
  Default: 2 (1 link-local + 1 global address).  
```

leandrolanzieri commented:
@aabadie you can search for a symbol using `/` and entering the name

aabadie commented:
And please squash!

leandrolanzieri commented:
@aabadie squashed. Thanks for reviewing!
